### ISSUE TITLE
[meta] Build and deploy single (megazord) multi-Rust component libraries

### ISSUE THREAD

**Issue by ncalexan** [mozilla/application-services#263]
This is a meta ticket tracking the effort to build and deploy single (megazord, or megazorded) libraries comprising multiple Rust component libraries.  For example, for Lockbox for Android, a single `.so`/AAR comprising the `fxa-client` and the `logins` Rust components.

There are a number of pieces I can think of that are relevant to this effort:

[ ] get automation to build and deploy combined Rust libraries
- [ ] get symbol visibility correct, pushing something like #168 (or an alternate approach)
- [ ] address end-to-end build times.  Spitball target: `git push` to built megazord in 45 minutes.
[ ] work out how the Android component wrappers can target megazord libraries in a uniform manner
[ ] perhaps ease the consuming application megazord consumption path with a Gradle plugin

I'm planning to track this work in the `a-s` repository initially because I expect most of the automation work to happen in this repository.

/cc @rfk @thomcc @vladikoff @csadilek

**Comment by ncalexan** [comment]
From @thomcc in Slack:
```
@nalexander re: megazording; if you decide to go the 'parsing' route and haven't come across it, https://github.com/Sean1708/rusty-cheddar looked like the simplest of the 'parse the rust crates and find their exports' that i could find. for the 'binary analysis' route, https://github.com/thomcc/application-services/blob/as-megazord/megazord-symbol-checker/src/main.rs could almost do enough to output the megazorded lib.rs on it's own (in particular you'd use the symbols from `components` to come up with what should be in it. i'm pretty sure you could do this as part of a build.rs, since writing build.rs's that output rust files is supported (edited)

the big caveat is that it will contain both c and rust symbols. there are a few ways to avoid this though (passing -fvisibility=hidden when compiling the c code, assuming the c code is the stuff that's in the intersection of the two libraries... probably other stuff too that I'm not thinking of)
```

**Comment by ncalexan** [comment]
A little status update here.  I've been pushing this along on a few fronts, mainly addressing:

> [ ] address end-to-end build times. Spitball target: git push to built megazord in 45 minutes.

[Bug 1502108](https://bugzilla.mozilla.org/show_bug.cgi?id=1502108) introduced a new `application-services-r` TC worker type, which should help our build times and build consistency.  (The WIP pull at #313 uses that worker type; I haven't bothered to use it in the current `.taskcluster.yml` yet.)

The WIP pull at #313 splits the build task into a (cached) libs build that compiles OpenSSL and SQLcipher in a separate task before building the Rust pieces.  I'm hoping that [Bug 1501720](https://bugzilla.mozilla.org/show_bug.cgi?id=1501720) will unblock doing the Docker parts of that work, but if it doesn't, I'll land some version of #313 faster in order to get this nice improvement.

I have WIP on building the libs for Desktop environments as well.  Linux is fine (as above), but macOS is a frustrating mess.  AFAICT OpenSSL is building but I can't quite coax SQLcipher across the line, yet.

The next steps:
- use sccache when compiling Rust (for all architectures)
- split `fxaclient` and `logins` (and eventually `places`) tasks
- share sccache cache across tasks

I have local WIP on splitting architectures in the [rust-android-gradle plugin](https://github.com/ncalexan/rust-android-gradle).  This has gotten hung up on trying to test the changes, which are getting to be non-trivial.  Testing Gradle plugins is frustrating.  It's not clear that this support is critical: if the sccache setup makes it fast to build each sub-project, even with all architectures, then we can simplify our lives.

I haven't seriously tried to move our publishing to `maven.mozilla.org`, although that will help a lot too.

**Comment by ncalexan** [comment]
#387 does everything here save for:

[ ] work out how the Android component wrappers can target megazord libraries in a uniform manner
[ ] perhaps ease the consuming application megazord consumption path with a Gradle plugin

That work spans repositories.  It will require some new idea (to achieve transparently, say with a Gradle plugin) or some small configuration in each consumer.
